chain/client/src/chunks_delay_tracker.rs

```diff
@@ -1,192 +1,103 @@
-use std::cmp::max;
-use std::collections::{BTreeMap, HashMap};
-use std::time::{Duration, Instant};
-
-use near_primitives::types::{BlockHeight, ShardId};
+use near_primitives::block::Block;
+use near_primitives::hash::CryptoHash;
+use near_primitives::sharding::ChunkHash;
+use std::collections::HashMap;
+use std::time::Instant;
 
 use crate::metrics;
 
-/// Provides monitoring information about the delays of receiving blocks and their corresponding chunks.
-/// Keeps timestamps of a limited number of blocks before the current head.
-/// Keeps timestamps of all blocks and chunks from the future.
-/// Doesn't make any assumptions about shard layout or the number of shards. If the head was moved forward, we assume that all requirements were satisfied.
-/// This object provides two metrics:
-/// 1) max delay between discovering a block and receiving the last chunk from that block.
-/// 2) the difference between the latest known block and the current head.
-/// If a chunk or a block is received multiple times, only the first time is recorded.
+/// Provides monitoring information about the delays of receiving and requesting blocks and their corresponding chunks.
+/// Keeps timestamps for all future blocks and chunks that were requested or received.
+/// Updates the metrics when the head gets updated. Assumes that if the head moves forward, all needed chunks were received.
+/// If a chunk or a block is received or requested multiple times, only the first time is recorded.
 #[derive(Debug, Default)]
 pub(crate) struct ChunksDelayTracker {
-    heights: BTreeMap<BlockHeight, HeightInfo>,
+    /// Timestamp of trying to process a block for the first time.
+    blocks_received: HashMap<CryptoHash, Instant>,
+    /// An entry gets created when a chunk gets requested for the first time.
+    /// Chunks get deleted when the block gets processed.
+    chunks_in_progress: HashMap<ChunkHash, ChunkInProgress>,
 }
 
-const CHUNKS_DELAY_TRACKER_HORIZON: u64 = 10;
-
-#[derive(Debug, Default)]
-struct HeightInfo {
-    block_received: Option<Instant>,
-    chunks_received: HashMap<ShardId, Instant>,
+#[derive(Debug)]
+/// Records timestamps of requesting and receiving a chunk. Assumes that each chunk is requested
+/// before it is received.
+struct ChunkInProgress {
+    /// Timestamps of requesting missing chunks.
+    chunk_requested: Instant,
+    /// Timestamps of receiving each of the shards in a block.
```

Comment:
nit: you say 'timestamps' - but there is a single timestamp, right?

Reply:
and is it 'each of the shards"? as I see only a single 'option<instant>'?

Reply:
Done